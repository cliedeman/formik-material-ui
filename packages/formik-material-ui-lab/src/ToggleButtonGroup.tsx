import * as React from 'react';
import MuiToggleButtonGroup, {
  ToggleButtonGroupProps as MuiToggleButtonGroupProps,
} from '@material-ui/lab/ToggleButtonGroup';
// import ToggleButton from '@material-ui/lab/ToggleButton';
import { FieldProps } from 'formik';
import invariant from 'tiny-warning';

export interface ToggleButtonGroupProps
  extends FieldProps,
    Omit<MuiToggleButtonGroupProps, 'name' | 'value' | 'error'> {
  type?: string;
}

export function fieldToToggleButtonGroup({
  field,
  type,
  onChange,
  onBlur,
  form,
  ...props
}: ToggleButtonGroupProps): MuiToggleButtonGroupProps {
  if (process.env.NODE_ENV !== 'production') {
    invariant(
      type === 'checkbox',
      `property type=checkbox is missing from field ${field.name}, this can caused unexpected behaviour`
    );

    if (!props.exclusive) {
      invariant(
        Array.isArray(field.value),
        `value for ${field.name} is not an array, this can caused unexpected behaviour`
      );
    }
  }

  const { onChange: _onChange, onBlur: _onBlur, ...fieldSubselection } = field;

  return {
    onBlur: onBlur
      ? onBlur
      : (event: React.FocusEvent<unknown>) => {
          field.onBlur(event ?? field.name);
        },
    onChange: onChange
      ? onChange
      : (_event: React.MouseEvent<HTMLElement>, newValue: unknown) => {
          form.setFieldValue(field.name, newValue);
        },
    ...fieldSubselection,
    ...props,
  };
}

export function ToggleButtonGroup(props: ToggleButtonGroupProps) {
  return <MuiToggleButtonGroup {...fieldToToggleButtonGroup(props)} />;
}

ToggleButtonGroup.displayName = 'FormikMaterialUIToggleButtonGroup';
