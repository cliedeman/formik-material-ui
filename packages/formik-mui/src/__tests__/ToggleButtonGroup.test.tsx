import React from 'react';
import { test, expect, vi } from 'vitest';
import { Field } from 'formik';

import { render } from './utils';

import { ToggleButtonGroup } from '../ToggleButtonGroup';

test('renders', async () => {
  const onSubmit = vi.fn();
  const { asFragment } = render(
    <Field component={ToggleButtonGroup} name="toggle" type="checkbox" />,
    {
      initialValues: {
        toggle: [],
      },
      onSubmit,
    }
  );

  expect(asFragment()).toMatchSnapshot();
});
