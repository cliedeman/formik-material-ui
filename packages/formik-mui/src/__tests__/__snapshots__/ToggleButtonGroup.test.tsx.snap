// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders 1`] = `
<DocumentFragment>
  <form
    action="#"
    data-testid="form"
  >
    <div
      class="MuiToggleButtonGroup-root css-bghv3y-MuiToggleButtonGroup-root"
      name="toggle"
      role="group"
    />
    <button
      data-testid="submit"
    >
      submit
    </button>
  </form>
</DocumentFragment>
`;
